metrics/xrootd_cache_eviction.go

```diff
@@ -0,0 +1,262 @@
+	}, []string{"dir_name", "type"})
+
+	XrootdCacheEvictionDirLastAccessTime = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "xrootd_cache_eviction_dir_last_access_time",
+		Help: "Last access time per directory",
+	}, []string{"dir_name", "type"})
+	XrootdCacheEvictionDirStBlocksUsage = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "xrootd_cache_eviction_dir_st_blocks_usage",
+		Help: "Storage blocks usage per directory",
+	}, []string{"dir_name"})
+	XrootdCacheEvictionDirNFilesOpen = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "xrootd_cache_eviction_dir_n_files_open",
+		Help: "Number of open files per directory",
+	}, []string{"dir_name"})
+	XrootdCacheEvictionDirNFiles = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "xrootd_cache_eviction_dir_n_files",
+		Help: "Number of files per directory",
+	}, []string{"dir_name"})
+	XrootdCacheEvictionDirNDirectories = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "xrootd_cache_eviction_dir_n_directories",
+		Help: "Number of directories per directory",
+	}, []string{"dir_name"})
+)
+
+func LaunchXrootdCacheEvictionMonitoring(ctx context.Context, egrp *errgroup.Group) {
+	egrp.Go(func() error {
+
+		statsFile := filepath.Join(param.Cache_StorageLocation.GetString(), "namespace", "pfc-stats", "DirStat.json")
+
+		interval := param.Xrootd_EvictionMonitoringInterval.GetInt()
+		ticker := time.NewTicker(time.Duration(interval) * time.Second)
+
+		// We use this to detect if the stats file is stale
+		var lastUpdateTime int
+
+		for {
+			select {
+			case <-ctx.Done():
+				log.Info("Xrootd cache eviction monitoring: context done")
+				return nil
+			case <-ticker.C:
+				log.Info("Xrootd cache eviction monitoring: tick")
+				stats, err := os.ReadFile(statsFile)
+				if err != nil {
+					log.Errorf("Xrootd cache eviction monitoring: failed to read stats file: %v", err)
+				}
+				var dirStatistics DirStatistics
+				err = json.Unmarshal(stats, &dirStatistics)
+				if err != nil {
+					log.Errorf("Xrootd cache eviction monitoring: failed to unmarshal stats file: %v", err)
+				}
+
+				log.Infof("Xrootd cache eviction monitoring: dirStatistics: %+v", dirStatistics)
+
+				if dirStatistics.UsageUpdateTime <= lastUpdateTime {
+					log.Infof("Xrootd cache eviction monitoring: directory stats is stale")
+					continue
+				}
+
+				lastUpdateTime = dirStatistics.UsageUpdateTime
+				XrootdCacheEvictionLastUpdateTime.Set(float64(lastUpdateTime))
+
+				XrootdCacheEvictionDiskUsage.Set(float64(dirStatistics.DiskUsed))
+
+				XrootdCacheEvictionSnapshotStatsResetTime.Set(float64(dirStatistics.SnapshotStatsResetTime))
+				XrootdCacheEvictionDiskTotal.Set(float64(dirStatistics.DiskTotal))
+				XrootdCacheEvictionFileUsage.Set(float64(dirStatistics.FileUsage))
+				XrootdCacheEvictionMetaTotal.Set(float64(dirStatistics.MetaTotal))
+				XrootdCacheEvictionMetaUsed.Set(float64(dirStatistics.MetaUsed))
+
+				for i, dirState := range dirStatistics.DirStates {
```

Comment:
This loop is worth splitting into a separate function so you can write a few unit tests for it.